### ISSUE TITLE
kubectl_manifest returning "Provider produced inconsistent result after apply" in Kubernetes 1.20

### ISSUE THREAD

**Issue by sculley** [gavinbunney/terraform-provider-kubectl#69]
## Description
I am not able to apply ```kubectl_manifest``` to Kubernetes 1.20, I receive an error. Although I can confirm the resources are created, the plan will exit with an error breaking the execution. Applying the same configuration in Kubernetes 1.18/1.19 works without returning any errors.

## Steps to reproduce

Whilst trying to apply the following

```hcl
resource "helm_release" "cert_manager" {
  name       = "cert-manager"
  repository = "https://charts.jetstack.io"
  chart      = "cert-manager"

  namespace  = "cert-manager"
  create_namespace = true

  set {
    name  = "installCRDs"
    value = "true"
  }
}

resource "kubernetes_secret" "cloudflare_token_api" {
  metadata {
    name = "cloudflare-token"
    namespace = "cert-manager"
  }

  data = {
    api-token = "secret_token"
  }

  depends_on = [
    helm_release.cert_manager
  ]
}

resource "kubectl_manifest" "issuer" {
  yaml_body = <<YAML
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: test
spec:
  acme:
    email: your@email.com
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: cloudflare-acme-private-key
    solvers:
    - dns01:
        cloudflare:
          email: your@email.com
          apiTokenSecretRef:
            name: cloudflare-token
            key: api-token
YAML

  depends_on = [
    helm_release.cert_manager,
    kubernetes_secret.cloudflare_token_api
  ]
}
```

I receive the error ```Root resource was present, but now absent.```

```
module.cert-manager.helm_release.cert_manager: Creating...
module.cert-manager.helm_release.cert_manager: Still creating... [10s elapsed]
module.cert-manager.helm_release.cert_manager: Still creating... [20s elapsed]
module.cert-manager.helm_release.cert_manager: Creation complete after 27s [id=cert-manager]
module.cert-manager.kubernetes_secret.cloudflare_token_api: Creating...
module.cert-manager.kubernetes_secret.cloudflare_token_api: Creation complete after 1s [id=cert-manager/cloudflare-api-token]
module.cert-manager.kubectl_manifest.issuer: Creating...

Error: Provider produced inconsistent result after apply

When applying changes to module.cert-manager.kubectl_manifest.issuer, provider
"registry.terraform.io/gavinbunney/kubectl" produced an unexpected new value:
Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

This is only occurring for me on Kubernetes 1.20, I have tested in 1.18/1.19 and it works

```
module.cert-manager.helm_release.cert_manager: Creating...
module.cert-manager.helm_release.cert_manager: Still creating... [10s elapsed]
module.cert-manager.helm_release.cert_manager: Still creating... [20s elapsed]
module.cert-manager.helm_release.cert_manager: Still creating... [30s elapsed]
module.cert-manager.helm_release.cert_manager: Still creating... [40s elapsed]
module.cert-manager.helm_release.cert_manager: Still creating... [50s elapsed]
module.cert-manager.helm_release.cert_manager: Creation complete after 57s [id=cert-manager]
module.cert-manager.kubernetes_secret.cloudflare_token_api: Creating...
module.cert-manager.kubernetes_secret.cloudflare_token_api: Creation complete after 1s [id=cert-manager/cloudflare-api-token]
module.cert-manager.kubectl_manifest.issuer: Creating...
module.cert-manager.kubectl_manifest.issuer: Creation complete after 2s [id=/apis/cert-manager.io/v1/namespaces/default/issuers/test]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

**Comment by debovema** [comment]
I have the same issue.
The only difference was that my original Kubernetes manifests were created in a 1.19 cluster which was then upgraded to 1.20
I thought the bug was caused by the upgrade but the root cause seems to be Kubernetes 1.20 itself.

**Comment by neomantra** [comment]
I think that the issue is that `selfLink` propagation is [disabled in K8S 1.20](https://github.com/kubernetes/enhancements/issues/1164) and will be removed in 1.21.

I ran `TF_LOG=trace terraform apply` and saw this line (timestamps removed for clarity:
```
[DEBUG] plugin.terraform-provider-kubectl_v1.9.4: [INFO] myns/myresource perform apply of manifest
[DEBUG] plugin.terraform-provider-kubectl_v1.9.4: [INFO] myns/myresource manifest applied, fetch resource from kubernetes
[DEBUG] plugin.terraform-provider-kubectl_v1.9.4: [DEBUG] myns/myresource fetched successfully, set id to: 
[DEBUG] plugin.terraform-provider-kubectl_v1.9.4: [DEBUG] myns/myresource fetch from kubernetes
[DEBUG] kubectl_manifest.test[0]: apply errored, but we're indicating that via the Error pointer rather than returning it: Provider produced inconsistent result after apply: When applying changes to kubectl_manifest.test[0], provider "registry.terraform.io/gavinbunney/kubectl" produced an unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

See the line `fetched successfully, set id to: ` ....
That corresponds to [this source line](https://github.com/gavinbunney/terraform-provider-kubectl/blob/master/kubernetes/resource_kubectl_manifest.go#L525):
```
d.SetId(response.GetSelfLink())
log.Printf("[DEBUG] %v fetched successfully, set id to: %v", manifest, d.Id())
```

So this provider is tracking resources with `selfLink`s, but K8S 1.20 doesn't have them anymore.  These IDs need to be extracted / generated some other way going forward.

**Comment by cristianburca** [comment]
I have the same issue 

**Server Version: v1.18.9-eks-d1db3c**

**main.tf**
`resource "kubectl_manifest" "keystore" {
  yaml_body = templatefile("${path.module}/templates/secret.yaml.tmpl", {
                name      = "keystore-secret"
                namespace = var.namespace
                type      = "Opaque"
                data      = yamlencode({
                              data = {
                                "keystore.p12"      = local.keystore
                                "keystore.password" = local.keystore_pass
                              }
                            })
              })
}
`

**secret.yaml.tmpl** 
`
apiVersion: v1
${data}
kind: Secret
metadata:
  name: ${name}
  namespace: ${namespace}
type: ${type}`

Error: Provider produced inconsistent final plan

```
When expanding the plan for
module.msk_secrets["strimzi-operator"].kubectl_manifest.keystore to include
new values learned so far during apply, provider
"registry.terraform.io/gavinbunney/kubectl" produced an invalid new value for
.yaml_body: inconsistent values for sensitive attribute.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

**Comment by Dr-Shadow** [comment]
@neomantra I'm trying to use your changes but I don't know how to install the binary once the code is built.
Can you give me directions about that ? Thank you in advance

**Comment by neomantra** [comment]
@Dr-Shadow 
I'm not sure if this is the best way, but it worked for me for now.... on Terraform 0.14.

Basically we fake a local 1.9.5 version (vs current published 1.9.4) .  The `neomantra.com` parts below don't actually hit anything external, it's just needed for the Terraform 0.14 provider loader naming conventions.

You need a Golang dev environment installed with a valid `GOPATH` ... then:
```
# clone and build my branch
git clone https://github.com/neomantra/terraform-provider-kubectl.git
cd terraform-provider-kubectl 
git checkout nm-69-selfLink
make build

# you can find the provider here:
ls -l $GOPATH/bin/terraform-provider-kubectl

# install it locally (assumes OSX, darwin_amd64, other platforms have other names/paths)
mkdir -p ~/.terraform.d/plugins/neomantra.com/terraform/kubectl/1.9.5/darwin_amd64/
cp $GOPATH/bin/terraform-provider-kubectl ~/.terraform.d/plugins/neomantra.com/terraform/kubectl/1.9.5/darwin_amd64/
```

Modify your `required_providers` stanza like so:
```
terraform {
  required_providers {
    kubernetes = {
      source = "hashicorp/kubernetes"
    }
    kubectl = {
      source  = "neomantra.com/terraform/kubectl"  // was "gavinbunney/kubectl"
      version = "= 1.9.5"
    }
  }
  required_version = ">= 0.14"
}
```

**Comment by ruckc** [comment]
i was able to do the go build and use my ~/.terraform.rc with
```
provider_installation {
  dev_overrides {
    "gavinbunney/kubectl" = "/home/ruckc/go/bin"
  }
}
```

but still get the same error:

```
kubectl_manifest.flannel: Creating...

Error: Provider produced inconsistent result after apply

When applying changes to kubectl_manifest.flannel, provider
"registry.terraform.io/gavinbunney/kubectl" produced an unexpected new value:
Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

**Comment by neomantra** [comment]
@ruckc  If you are building from my PR, you can add some extra printing to verify you are make it easier to see -- even as a sanity check to ensure it actually is using the patched provider.     

I'd be curious what the value of the `selfLink` is, because if it is empty (length 0), a new link should be generated by this patch.    Maybe something else (whitespace?) is getting returned on your K8S instance and I need to add a different check.

**Comment by ruckc** [comment]
I'm just trying to apply the flannel.yaml to a new kubeadm cluster...

attached is the full trace
[kubectl.txt](https://github.com/gavinbunney/terraform-provider-kubectl/files/5852143/kubectl.txt)

**Comment by neomantra** [comment]
I see this line in your log -- I don't think it is using the patched provider:

```
2021-01-21T20:38:15.449Z [DEBUG] plugin: plugin process exited: path=.terraform/providers/registry.terraform.io/gavinbunney/kubectl/1.9.4/linux_arm64/terraform-provider-kubectl_v1.9.4 pid=42354
```

**Comment by ruckc** [comment]
okay, i had my .terraformrc named wrong, and the content was wrong.

```hcl
provider_installation {
  dev_overrides {
    "gavinbunney/kubectl" = "/home/ruckc/go/bin"
  }
  direct {

  }
}
```

this let me skip through, and it works just dandy with this tf file:

```hcl
data "http" "flannel" {
  url = "https://raw.githubusercontent.com/coreos/flannel/master/Documentation/kube-flannel.yml"
}

data "kubectl_file_documents" "flannel_manifests" {
  content = data.http.flannel.body
}

resource "kubectl_manifest" "flannel" {
  count     = length(data.kubectl_file_documents.flannel_manifests.documents)
  yaml_body = element(data.kubectl_file_documents.flannel_manifests.documents, count.index)
}
```

**Comment by linuxbsdfreak** [comment]
Hi @gavinbunney ,

Same issue on 1.20

```
terraform {
  required_version = ">= 0.14"
  required_providers {
    kubectl = {
      source = "gavinbunney/kubectl"
      version = "1.9.4"
    }
    flux = {
      source  = "fluxcd/flux"
      version = "0.0.11"
    }
  }
}
provider "flux" {
  alias = "gitops"
}
provider "kubectl" {
  alias = "shoot"
}
provider "kubernetes" {
  alias = "shoot"
}
provider "github" {
  alias = "gitops"
}
data "github_repository" "gitops-demo" {
  provider = github.gitops
  full_name = join("/", [ var.github_org, var.repository_name])
}
# Flux
data "flux_install" "master" {
  provider = flux.gitops
  target_path = var.target_path
}
data "flux_sync" "master" {
  provider = flux.gitops
  target_path = var.target_path
  url         = data.github_repository.gitops-demo.http_clone_url
  branch      = var.branch
}
# Kubernetes
resource "kubernetes_namespace" "flux_system" {
  provider = kubernetes.shoot
  metadata {
    name = "flux-system"
  }
  lifecycle {
    ignore_changes = [
      metadata[0].labels,
    ]
  }
}
data "kubectl_file_documents" "install" {
  provider = kubectl.shoot
  content = data.flux_install.master.content
}
data "kubectl_file_documents" "sync" {
  provider = kubectl.shoot
  content = data.flux_sync.master.content
}
locals {
  install = [ for v in data.kubectl_file_documents.install.documents : {
      data: yamldecode(v)
      content: v
    }
  ]
  sync = [ for v in data.kubectl_file_documents.sync.documents : {
      data: yamldecode(v)
      content: v
    }
  ]
}
resource "kubectl_manifest" "install" {
  provider = kubectl.shoot
  for_each   = { for v in local.install : lower(join("/", compact([v.data.apiVersion, v.data.kind, lookup(v.data.metadata, "namespace", ""), v.data.metadata.name]))) => v.content }
  depends_on = [kubernetes_namespace.flux_system]
  yaml_body = each.value
}
resource "kubectl_manifest" "sync" {
  provider = kubectl.shoot
  for_each   = { for v in local.sync : lower(join("/", compact([v.data.apiVersion, v.data.kind, lookup(v.data.metadata, "namespace", ""), v.data.metadata.name]))) => v.content }
  #depends_on = [kubectl_manifest.install, kubernetes_namespace.flux_system]
  depends_on = [kubernetes_namespace.flux_system]
  yaml_body = each.value
}
resource "kubernetes_secret" "master" {
  provider = kubernetes.shoot
  depends_on = [kubectl_manifest.install]
  metadata {
    name      = data.flux_sync.master.namespace
    namespace = data.flux_sync.master.namespace
  }
  data = {
    password   = var.github_token
    username   = var.github_owner
  }
}
# Github
resource "github_repository_file" "install" {
  provider = github.gitops
  repository = var.repository_name
  file       = data.flux_install.master.path
  content    = data.flux_install.master.content
  branch     = var.branch
}
resource "github_repository_file" "sync" {
  provider = github.gitops
  repository = var.repository_name
  file       = data.flux_sync.master.path
  content    = data.flux_sync.master.content
  branch     = var.branch
}
resource "github_repository_file" "kustomize" {
  provider = github.gitops
  repository = var.repository_name
  file       = data.flux_sync.master.kustomize_path
  content    = data.flux_sync.master.kustomize_content
  branch     = var.branch
}
output "gitops-demo-repo-url" {
  value = data.github_repository.gitops-demo.http_clone_url
  description = "gitops demo repo url"
}
```

**Comment by imcitius** [comment]
@gavinbunney same here on 1.20.

**Comment by fsalum** [comment]
is @gavinbunney still maintaining this provider ? there is a PR open to fix the issue and no updates from him

**Comment by scoutmindset18** [comment]
Same issue here on 1.20, I see a PR 24 days old open to fix the issue without any comment or review. is @gavinbunney active maintaining this ?

**Comment by gaquino** [comment]
Same issue here on 1.20. I see a PR 24 days old open to fix the issue without any comment or review. is @gavinbunney active in maintaining this?

**Comment by gavinbunney** [comment]
Released as v1.10
